Configuration Management/Config Distro Integration

tadean opened the pull request:
This pull request adds the `config_sync` / `config_distro` ecosystem to quickstart and creates a `config_provider` plugin that offers similar behavior to our existing system of handling overrides to contrib module configuration.

## Description
This pull request adds the `config_sync` / `config_distro` ecosystem to quickstart and creates a `config_provider` plugin that offers similar behavior to our existing system of handling overrides to contrib module configuration; this operates on a `config/quickstart` folder and has the effect of adding the configuration override change to whichever module originally owned the configuration.

For example, if the `cas` module originally deployed the `cas.settings` configuration, the additional `config_provider` plugin will make `az_cas`'s version of `config/quickstart/cas.settings.yml` visible to `config_sync` as an update to the `cas`modules settings, in other words allowing our overrides to participate in the `config_distro`ecosystem.

This creates the issue that a new install of quickstart has several pending distribution updates out of the gate (e.g., `cas` showing the updates from `az_cas`'s being enabled. This is resolved by doing an automatic import of the `config/quickstart` overrides when the module containing them is first enabled. This means, for example, that `az_cas`'s generated updated to `cas` config is automatically imported when `az_cas` is first enabled.

`config_sync` is D9 ready, but has one issue with the core version requirement on a mostly-stub testing submodule it contains. See my Drupal.org issue about this [here](https://www.drupal.org/project/config_sync/issues/3176955). This is a patch against an .info file, so I also have a gist [here](https://gist.github.com/tadean/602c6412fbbdcc16a2d8a0fb58642f9e) for the current packaged version that takes into account the drupal.org autopackaging additions.

A new unit test is added to verify that configuration overrides work correctly.

## Related Issue
#175 

## How Has This Been Tested?

- Verify that existing `az_cas` settings are imported.
- Make local changes to the embedded configuration of quickstart distribution modules, e.g. `az_flexible_page`, `az_paragraphs`, etc.
- Verify that the file changes are visible as distribution updates at `/admin/config/development/distro`

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I've added this Pull Request to the AZ Quickstart project in the right column.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

tadean commented:
One issue pending resolution:

When enabling a module that contains configuration overrides, (e.g. `az_cas`), the changes will be correctly applied, but the `config_sync` snapshot of the module is not updated. This results in behavior where `config_distro` sees an empty "update" for the module that contains zero changes:

![image](https://user-images.githubusercontent.com/51835841/101659154-f8982c00-3a02-11eb-8b0e-f95c99ada954.png)

To resolve this, the `config_sync` snapshot should be updated after the overrides are applied.

joeparsons commented:
Do you think we need to look into these errors identified by PHPstan before merging?
```
 ------ ----------------------------------------------------------------------------------------------------------- 
  Line   modules/contrib/config_distro/modules/config_distro_ignore/src/Plugin/ConfigFilter/DistroIgnoreFilter.php  
 ------ ----------------------------------------------------------------------------------------------------------- 
  138    Calls to function fnmatch should not exist.                                                                
  158    Calls to function fnmatch should not exist.                                                                
 ------ ----------------------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------- 
  Line   modules/contrib/config_update/config_update_ui/src/Commands/ConfigUpdateUiCommands.php  
 ------ ---------------------------------------------------------------------------------------- 
  31     Call to deprecated method logger() of class Drush\Drush:                                
         Use injected logger instead.                                                            
 ------ ---------------------------------------------------------------------------------------- 

 -- --------------------------------------------------------------------------- 
     Error                                                                      
 -- --------------------------------------------------------------------------- 
     Internal error: Internal error: Invalid value in file                      
     /src/az-quickstart-scaffolding/web/modules/contrib/config_update/config_u  
     pdate_ui/src/Logger/ConfigUpdateUiDrush8Logger.php                         
     Run PHPStan with --debug option and post the stack trace to:               
     https://github.com/phpstan/phpstan/issues/new                              
     Internal error: Internal error: Invalid value in file                      
     /src/az-quickstart-scaffolding/web/modules/contrib/config_update/config_u  
     pdate_ui/config_update_ui.drush_testing.inc                                
     Run PHPStan with --debug option and post the stack trace to:               
     https://github.com/phpstan/phpstan/issues/new                              
 -- --------------------------------------------------------------------------- 

 [ERROR] Found 5 errors  
```

tadean commented:
> Do you think we need to look into these errors identified by PHPstan before merging?
> 
> ```
>  ------ ----------------------------------------------------------------------------------------------------------- 
>   Line   modules/contrib/config_distro/modules/config_distro_ignore/src/Plugin/ConfigFilter/DistroIgnoreFilter.php  
>  ------ ----------------------------------------------------------------------------------------------------------- 
>   138    Calls to function fnmatch should not exist.                                                                
>   158    Calls to function fnmatch should not exist.                                                                
>  ------ ----------------------------------------------------------------------------------------------------------- 
> 
>  ------ ---------------------------------------------------------------------------------------- 
>   Line   modules/contrib/config_update/config_update_ui/src/Commands/ConfigUpdateUiCommands.php  
>  ------ ---------------------------------------------------------------------------------------- 
>   31     Call to deprecated method logger() of class Drush\Drush:                                
>          Use injected logger instead.                                                            
>  ------ ---------------------------------------------------------------------------------------- 
> 
>  -- --------------------------------------------------------------------------- 
>      Error                                                                      
>  -- --------------------------------------------------------------------------- 
>      Internal error: Internal error: Invalid value in file                      
>      /src/az-quickstart-scaffolding/web/modules/contrib/config_update/config_u  
>      pdate_ui/src/Logger/ConfigUpdateUiDrush8Logger.php                         
>      Run PHPStan with --debug option and post the stack trace to:               
>      https://github.com/phpstan/phpstan/issues/new                              
>      Internal error: Internal error: Invalid value in file                      
>      /src/az-quickstart-scaffolding/web/modules/contrib/config_update/config_u  
>      pdate_ui/config_update_ui.drush_testing.inc                                
>      Run PHPStan with --debug option and post the stack trace to:               
>      https://github.com/phpstan/phpstan/issues/new                              
>  -- --------------------------------------------------------------------------- 
> 
>  [ERROR] Found 5 errors  
> ```

I think it would be good to. I believe there is already an open issue in `config_distro` about the fnmatch issue.

tadean commented on modules/custom/az_core/src/AZConfigOverride.php:
```diff
@@ -0,0 +1,80 @@
+use Drupal\az_core\Plugin\ConfigProvider\QuickstartConfigProvider;
+
+/**
+ * Class AZConfigOverrideIMport.
```
This comment should be updated to provide better documentation of the override service.

trackleft reviewed (dismissed):
Tested locally, thanks!

trackleft commented:
Found the patch troy mentioned here: https://www.drupal.org/project/config_distro/issues/3144145

Also found this patch https://www.drupal.org/files/issues/2020-04-03/3089420-8.patch for the logger error

tadean commented:
> Found the patch troy mentioned here: https://www.drupal.org/project/config_distro/issues/3144145
> 
> Also found this patch https://www.drupal.org/files/issues/2020-04-03/3089420-8.patch for the logger error

Thanks!

tadean commented:
Noticed one issue in testing:

![image](https://user-images.githubusercontent.com/51835841/101840705-dafdbc00-3b01-11eb-8b0e-91b3d378c4b1.png)


When updates exist for multiple modules, and some are deselected (to not add the update), it appears to import the unselected ones anyway, e.g. the checkboxes don't seem to disable the imports.

This seems to happen on a clean non-quickstart Drupal installation too. This worked correctly in an earlier version, so it's probably a recent issue in one of the contrib modules. I'll see if I can locate the source of that.

tadean commented:
Added the patch to address the `fnmatch` phpstan issue.

Reviewing the functional status of the `config_update` issues as it looks like a few patches are necessary to fully address the issue and some are in a postponed state.

tadean commented:
- Added code to re-snapshop overridden `config_sync` config for override files
- Added in expected default config hash values
- Specified PHPStan exclusions for the `config_update` files that are undergoing Drush 9 refactor
